Kiali Tutorial

lucasponce opened the pull request:
~~Not yet ready.~~ 
It's in ~~30%~~ ~~60%~~ ~~80%~~ 95% of progress.

Okey, ~100 files PR with new content, it's ready for review.

Some considerations:

- Content is done in the staging area, accessing from main top "Tutorial" link will fail as it's not yet updated into the "latest" section, when content is reviewed it will be copied there (but I don't want to x2 resources yet).
- I'd apreciate if reviewers can follow the steps and feel like a user, not only review the wording, just to see if I miss some important configuration step, or logical step that it's implicit to us but important to the user to not miss the path.
- This is a first step, not all features are presented (yet), the idea was to link with https://istio.io/latest/docs/tasks/ to allow user to start discovering Istio ServiceMesh from Kiali, in that context, more deep features (config validation or similar) were not mention on purpose.
- Also, the structure may vary in the future, this is just a first approach in order to have minor steps documented about how to do something very specific, also to allow user to be more autonomous (specially on the management features).

Thanks in advance for the review !

Fixes https://github.com/kiali/kiali/issues/3488

xeviknal reviewed (commented):
I reviewed it til the 04-observe so far. I'd like to drop some early comments. I'll provide some more comments later.

xeviknal commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+NOTE: *[2]* Create Authorization Policies for current traffic for *travel-agency* namespace
+
+Every workload in the cluster users a link:https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/[Service Account].
```
users -> uses

jotak commented on content/documentation/staging/tutorial/01-prerequisites/03-update-kiali.adoc:
```diff
@@ -0,0 +1,29 @@
+---
+title: Update Kiali
+weight: 14
```
On my machine I see the section "Access the Kiali UI" before "Update Kiali", which I guess isn't intended. I can see they have the same weight, perhaps that's the reason.

lucasponce replied:
Ouch, yes, thanks good catch !

jotak commented on content/documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc:
```diff
@@ -0,0 +1,58 @@
+---
+title: Review the demo application
```
This chapter explains quite in details what is the travels demo, so perhaps it could have a more explicit title, something like "Understanding the demo application" or "Explaining the demo application" ?

jotak commented on content/documentation/staging/tutorial/01-prerequisites/02-install-istio.adoc:
```diff
@@ -0,0 +1,15 @@
+
+Once you have your Kubernetes cluster ready, follow the link:https://istio.io/latest/docs/setup/getting-started/[Istio Getting Started] to install and setup a demo profile that will be used in this tutorial.
+
+icon:bullhorn[size=1x]{nbsp} Determining ingress IP and ports and create DNS entries will be necessary in the following steps.
```
not sure if "both" is necessary but +1 for creating

jotak commented on content/documentation/staging/tutorial/03-first-steps/02-enable-sidecars.adoc:
```diff
@@ -0,0 +1,33 @@
+
+We are going to start with the *control* workload deployed into the *travel-control* namespace:
+
+NOTE: *[1]* Enable Auto Injection on *travel-control* namespace
```
Not sure if it's just me, but seeing numbers in brackets like this sound like footnotes, not steps. I'd suggest something like **1.** instead

lucasponce replied:
Perhaps using "[1] -> Step 1", to differentiate of a list.

jotak commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+</a>
+++++
+
+{nbsp} +
```
Something is missing here I guess. Perhaps: `Click on "Add Rule" so that any request will be routed to the *control* workload.`

jotak commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+NOTE: *[4]* Review *travel-control* namespace in Kiali
+++++
+<a class="image-popup-fit-height" href="/images/tutorial/03-03-travel-control-graph.png" title="Travel Control Graph">
+    <img src="/images/tutorial/03-03-travel-control-graph.png" style="display:block;margin: 0 auto;" />
```
The image is not 100% accurate as it shows destination services & workloads as if they were in the mesh, but they aren't yet at this point of the tutorial
Here's what I have at this step: 
![Capture d’écran de 2021-01-18 14-14-07](https://user-images.githubusercontent.com/2153442/104920080-8e768d80-5997-11eb-9d52-6955329b6db0.png)

It's not such a big problem, but perhaps is a bit misleading for the users.

jotak replied:
hmm perhaps you're right there's something wrong on my side. I'll double check

jshaughn replied:
I agree with this, at this point in the tutorial we have only instrumented the control workload.  So we should not have telemetry out of control.  Actually, for me I don't even have the requests to the travel-portal services.

lucasponce replied:
I've repeat the very same step several times in minikube platform, and now with Kiali v1.29.0 I have a similar image:
![image](https://user-images.githubusercontent.com/1662329/105716474-25ef5980-5f1f-11eb-875f-52051a9fb213.png)

So, I'm curious if the differences may be due timeline (did you let the travel control app creates enough traffic ?) or platform specific.

jotak replied:
I don't think it was a matter of insufficient traffic, I did let it run for minutes, perhaps tens of minutes.
I'm going to retry it perhaps tomorrow. But this isn't a blocker anyway. If we find something to add to the tutorial, we can still add it later.
As of now, the tutorial is in a very good shape!

jotak commented on content/documentation/staging/tutorial/04-observe/01-enable-sidecars-for-all-namespaces.adoc:
```diff
@@ -0,0 +1,59 @@
+
+Proxies connect with the control plane and provide link:https://istio.io/latest/docs/concepts/what-is-istio/#why-use-istio[Service Mesh functionality].
+
+One of the main features is the automatic metrics, logs and traces provide automatically by the sidecar.
```
yes, surely the second one can simply be removed

jotak commented on content/documentation/staging/tutorial/04-observe/01-enable-sidecars-for-all-namespaces.adoc:
```diff
@@ -0,0 +1,59 @@
+kubectl label namespace travel-agency istio-injection=enabled
+kubectl label namespace travel-portal istio-injection=enabled
+
+kubectl delete -f <(curl -L https://raw.githubusercontent.com/kiali/demos/master/travels/travel_agency.yaml) -n travel-agency
```
instead of removing / resending everything we could also just delete the pods, this is perhaps more straightforward:

```
kubectl delete pods --all -n travel-agency
kubectl delete pods --all -n travel-portal
```

lucasponce replied:
I'm going to test, but not sure if if works if the controller is not update the Istio webhook chain may not known that it needs to inject a new sidecar.

But I guess if you update the controller, then the webhook checks if namespace label is present.

lucasponce replied:
Yes, I'm afraid this mechanism works as I described.

```
1.- Anotate the namespace with the
2.- Create the controller
3.- The controller generates the pod with the sidecar
```

Existing controllers needs to be removed (or updated) but as annotation lives at namespace level, it doesn't help to just annotate the namespace + delete the pods, then the controller is not taking the changes from the Webhook chain.

jotak replied:
hmm i've always just deleted pods, and it's always worked for me (not just for this demo, but in general)

jotak commented:
Taking a break, I'm not even half-way! :-) I'll continue later.
But this is already an awesome piece of tutorial @lucasponce !

jmazzitelli commented:
The deploy failed - so we can't view the end result: https://app.netlify.com/sites/kiali/deploys/5ffdccf6bba99e0007611ad3
How are  people reviewing? Pulling the PR and building locally?

jmazzitelli reviewed (commented):
I pulled the PR and ran "make  serve" and got these 4 warnings at the top:
```
Building sites …
ERROR 2021/01/18 17:10:10 documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc: asciidoctor: WARNING: <stdin>: line 6: section title out of sequence: expected level 1, got level 4
ERROR 2021/01/18 17:10:10 documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc: asciidoctor: WARNING: <stdin>: line 16: section title out of sequence: expected level 1, got level 4
ERROR 2021/01/18 17:10:10 documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc: asciidoctor: WARNING: <stdin>: line 34: section title out of sequence: expected level 1, got level 4
ERROR 2021/01/18 17:10:10 documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc: asciidoctor: WARNING: <stdin>: line 43: section title out of sequence: expected level 1, got level 4
```

jshaughn commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+
+Then from this machine, the url `control.travel-control.istio-cluster.org` will resolve to the External IP of the Ingress Gateway of Istio.
+
+NOTE: *[2]* Use the Request Routing Wizard on *control* service to generate a traffic rule
```
on the *control*

jshaughn commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+{nbsp} +
+Understanding what happened:
+
+. Extenal traffic enters into the cluster through a Gateway
```
External

jshaughn commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+Understanding what happened:
+
+. Extenal traffic enters into the cluster through a Gateway
+. Traffic is routed to *control* service through a VirtualService
```
the *control*

Here you introduce VirtualService for the first time.  Perhaps above you should mention above what the wizard will create for you.

lucasponce replied:
I've added a reference to the VirtualService but the Wizard scenario and what generates is described in the "connect" step a couple of points later, if I introduce this here it triggers to add more context that is reserved to other section.

jshaughn commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+
+. Extenal traffic enters into the cluster through a Gateway
+. Traffic is routed to *control* service through a VirtualService
+. Kiali Graph visualizes the telemetry reported from *control* sidecar proxy
```
the traffic telemetry reported from the *control*

jshaughn commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+. Extenal traffic enters into the cluster through a Gateway
+. Traffic is routed to *control* service through a VirtualService
+. Kiali Graph visualizes the telemetry reported from *control* sidecar proxy
+.. Only *travel-control* namespace is part of the mesh
```
the *travel-control*

jshaughn reviewed (commented):
Releasing First Batch of Comments...

jshaughn commented on content/documentation/staging/tutorial/05-connect/01-request-routing.adoc:
```diff
@@ -0,0 +1,158 @@
+
+Once the Request Routing is working we can verify that outbound traffic from every portal goes all to a single *travels* workload.
+
+If you use the "Find.../Hide..." filter, graph can hide old traffic grey edges for a clear graph.
+
+++++
+<a class="image-popup-fit-height" href="/images/tutorial/05-01-request-routing-graph.png" title="Travels Portal Namespace Graph">
+    <img src="/images/tutorial/05-01-request-routing-graph.png" style="display:block;margin: 0 auto;" />
+</a>
+++++
```
This is no longer necessary as by default we no longer show "Idle Edges".  Only if you enable the "Show Idle Edges" display option is the `!traffic` Hide option useful.

lucasponce replied:
But I think it applies for the version 1.28.0 used for the tutorial.
We can change that later, once we consolidate it, we can review it when we publish new versions if some feature needs some adjustment.

jshaughn commented on content/documentation/staging/tutorial/05-connect/01-request-routing.adoc:
```diff
@@ -0,0 +1,158 @@
+++++
+
+{nbsp} +
+Using a large interval we can see how a workload received traffic from all portals and moved to a single portal once that the Request Routing scenarios was defined.
```
...the workload initially received traffic from all portals but then only a single portal after the Request Routing scenarios were defined.

jshaughn commented on content/documentation/staging/tutorial/05-connect/01-request-routing.adoc:
```diff
@@ -0,0 +1,158 @@
+
+NOTE: *[5]* Update or delete Istio Configuration
+
+Kiali Wizards allow to define high level Service Mesh scenarios and generate the Istio Configuration (VirtualServices, DestinationRules, Gateways and PeerRequests) needed for its implementation.
```
Kiali Wizards allow you to define a high level Service Mesh scenario and will then generate the Istio Configuration needed for its implementation (VirtualServices, DestinationRules, Gateways and PeerRequests).

jshaughn commented on content/documentation/staging/tutorial/05-connect/01-request-routing.adoc:
```diff
@@ -0,0 +1,158 @@
+
+Kiali Wizards allow to define high level Service Mesh scenarios and generate the Istio Configuration (VirtualServices, DestinationRules, Gateways and PeerRequests) needed for its implementation.
+
+These scenarios can be updated and the generated Istio configuration linked for a given service can be also deleted from the "Actions" menu.
```
These scenarios can be updated or deleted from the "Actions" menu of a given service.  The relevant Istio configuration will be modified or removed, as needed.

jshaughn commented on content/documentation/staging/tutorial/05-connect/01-request-routing.adoc:
```diff
@@ -0,0 +1,158 @@
+++++
+
+{nbsp} +
+As part of the tutorial step, when you have finished to experiment more Routing Requests scenarios, you can delete the generated Istio config for the *travels* service.
```
As part of this tutorial step, when you have finished experimenting with Routing Request scenarios then use the *travels* service "Actions" menu to delete the generated Istio config.

jshaughn reviewed (commented):
Good stuff, second batch of review comments.

jmazzitelli commented:
There are so many requested changes here already, I'll wait until they are incorporated and then do my review. That way I won't cover the same ground as everyone else did and I'll be able to review the latest update.

jshaughn commented on content/documentation/staging/tutorial/05-connect/03-traffic-shifting.adoc:
```diff
@@ -0,0 +1,60 @@
+
+This Traffic Shifting step will simulate a new scenario: the new *travels-v2* and *travels-v3* workloads will represent new improvements for the *travels* service that will be used by all requests.
+
+These new improvements implemented in *travels-v2* and *travels-v3* represent two alternatives way to address a specific problem and we would like to test them before to decide which one to use as a next version.
```
to decide -> deciding

jshaughn commented on content/documentation/staging/tutorial/05-connect/03-traffic-shifting.adoc:
```diff
@@ -0,0 +1,60 @@
+++++
+
+{nbsp} +
+Create a scenario with 80% of the traffic distributed on *travels-v1* workload and 10% of the traffic distributed on *travels-v2* and *travels-v3*.
```
distributed on --> distributed each on

jshaughn commented on content/documentation/staging/tutorial/05-connect/03-traffic-shifting.adoc:
```diff
@@ -0,0 +1,60 @@
+
+NOTE: *[3]* Compare *travels* workload and study new changes proposed in *travels-v2* and *travels-v3*
+
+Istio Telemetry is grouped per logical application and that gives an advantage to compare different workloads that are related to one or more services that represent a similar concept.
```
I found this sentence confusing, I think you are saying something like the following, but this may need more work:

Istio Telemetry is grouped per logical application.  That has the advantage of easily comparing different but related workloads, for one or more services.

jshaughn commented on content/documentation/staging/tutorial/05-connect/03-traffic-shifting.adoc:
```diff
@@ -0,0 +1,60 @@
+
+NOTE: *[4]* Update or delete Istio Configuration
+
+As part of this step you can update the Traffic Shifting scenario to test different distribution, when finished you can delete the generated Istio config for the *travels* service.
```
distribution, when   -->  distributions.  When

jshaughn commented on content/documentation/staging/tutorial/05-connect/04-tcp-traffic-shifting.adoc:
```diff
@@ -0,0 +1,74 @@
+
+The Travel Demo application has a database service used by several services deployed in the *travel-agency* namespace.
+
+At some point in the lifecycle of the application, the telemetry shows that the database service degrades and starts to increase the average response time.
```
no , needed

jshaughn commented on content/documentation/staging/tutorial/05-connect/04-tcp-traffic-shifting.adoc:
```diff
@@ -0,0 +1,74 @@
+
+Our database specialist is suggesting two approaches and proposes to prepare two versions of the database service to test which may work better.
+
+This step will show how the "Traffic Shifting" strategy can be applied to TCP services to test which database new indexes strategy works better.
```
database new indexes strategy  -->  new database indexing strategy

jshaughn commented on content/documentation/staging/tutorial/05-connect/04-tcp-traffic-shifting.adoc:
```diff
@@ -0,0 +1,74 @@
+
+NOTE: *[1]* Deploy *mysqldb-v2* and *mysqldb-v3* workloads
+
+To deploy the new versions of the *mysqldb* service follow the steps:
```
follow the steps --> execute the commands

jshaughn commented on content/documentation/staging/tutorial/05-connect/04-tcp-traffic-shifting.adoc:
```diff
@@ -0,0 +1,74 @@
+++++
+
+{nbsp} +
+Create a scenario with 80% of the traffic distributed on *mysqldb-v1* workload and 10% of the traffic distributed on *mysqldb-v2* and *mysqldb-v3*.
```
distributed on  -->  distributed each on

jshaughn commented on content/documentation/staging/tutorial/05-connect/04-tcp-traffic-shifting.adoc:
```diff
@@ -0,0 +1,74 @@
+
+NOTE: *[4]* Compare *mysqldb* workload and study new indexes proposed in *mysqldb-v2* and *mysqldb-v3*
+
+TCP services have different telemetry, but it's still grouped by versions allowing to compare and study differences on the patterns of *mysqldb-v2* and *mysqldb-v3*.
```
TCP services have different telemetry but it's still grouped by versions, allowing the user to compare and study pattern differences for *mysqldb-v2* and *mysqldb-v3*.

jshaughn commented on content/documentation/staging/tutorial/05-connect/04-tcp-traffic-shifting.adoc:
```diff
@@ -0,0 +1,74 @@
+++++
+
+{nbsp} +
+The charts show some peaks in *mysqldb-v2* compared with *mysqldb-v3* but overall a similar behaviour, so probably could be safe to choose any of the both proposals to shift all traffic.
```
behavior, so it's probably safe to choose either strategy to shift all traffic.

jshaughn commented on content/documentation/staging/tutorial/05-connect/04-tcp-traffic-shifting.adoc:
```diff
@@ -0,0 +1,74 @@
+
+NOTE: *[4]* Update or delete Istio Configuration
+
+As part of this step you can update the TCP Traffic Shifting scenario to test different distribution, when finished you can delete the generated Istio config for the *mysqldb* service.
```
Break into two sentences:   . When finished...

jshaughn reviewed (commented):
Batch 3, almost done!

lucasponce commented:
@jmazzitelli 

> How are people reviewing? Pulling the PR and building locally?

Yes, I just perform a "make serve" locally, Ignore some of the errors and review the tutorial from the staging area.

As it proposes a new first level menu, staging content needs to be copied to "latest" but to not x2 duplicaate the files, I'll do that once the content is more or less reviewed.

lucasponce commented:
Thanks for all the comments ! 
I'm going to work today on them.

lucasponce commented on content/documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc:
```diff
@@ -0,0 +1,58 @@
+
+++++
+<a class="image-popup-fit-height" href="/images/tutorial/02-02-travels-demo-design.png" title="Travels Demo Design">
+    <img src="/images/tutorial/02-02-travels-demo-design.png" style="display:block;margin: 0 auto;" />
```
Good catch, yeah, I missed that.

lucasponce commented on content/documentation/staging/tutorial/04-observe/01-enable-sidecars-for-all-namespaces.adoc:
```diff
@@ -0,0 +1,59 @@
+:icons: font
+:sectlinks:
+
+Istio sidecar proxy allows to add a workload into the mesh.
```
Thanks for the simplification, some time I think the phrase in a different language and then I translate, that's why I use these verbs so often.

lucasponce commented on content/documentation/staging/tutorial/04-observe/02-graph.adoc:
```diff
@@ -0,0 +1,49 @@
+{nbsp} +
+Review the status of the mesh, everything seem healthy, but also note that *hotels* service have more load compared with other services inlucded in the *travel-agency* namespace.
+
+NOTE: *[2]* Select *hotels* service and correlate with traces
```
> Note also that the side panel image is now dated as the trace detail has recently changed format:

Well, that's going to happen when we release new versions.
I used the latest version deployed in the helm, we'll need to review screenshots as an on-going task.

lucasponce commented on content/documentation/staging/tutorial/04-observe/03-details.adoc:
```diff
@@ -0,0 +1,103 @@
+:icons: font
+:sectlinks:
+
+Kiali provides link:../features/#_detail_views[Detail Views] to navigate into application, workload and services.
```
+1, I moved all the links to external tabs for consistency and to not miss the progress, even if user is moving back to a previous section.

lucasponce commented on content/documentation/staging/tutorial/05-connect/07-mirroring.adoc:
```diff
@@ -0,0 +1,75 @@
+The link:../tutorial/#03-traffic-shifting[Traffic Shifting] step was focused on *travels* service adding a new *travels-v2* and *travels-v3* workloads
+and the link:../tutorial/#04-tcp-traffic-shifting[TCP Traffic Shifting] showed how this scenario can be used on TCP services like *mysqldb* service.
+
+Mirroring (or shadowing) is a particular case of the Traffic Shifting scenario when the proxy sends a copy of live traffic to a mirrored service.
```
+1 on when -> where, the service -> service instead I don't see it, as it's not a replacement, but an addition.

lucasponce commented on content/documentation/staging/tutorial/05-connect/07-mirroring.adoc:
```diff
@@ -0,0 +1,75 @@
+
+Mirroring (or shadowing) is a particular case of the Traffic Shifting scenario when the proxy sends a copy of live traffic to a mirrored service.
+
+The mirrored traffic happens out of band of the request path for a primary service allowing to test new services in production environments with minimal risk.
```
The "out of band of the request path" expression is taken from [1].

It means that "mirrored" traffic is a copy, like the "tap" pattern in classic SOA.

Imagine 2 services: A -> B, and you have now a mirror called B-Mirror, so A request to B are copied to B-Mirror, but B-Mirror responses are not caught by A because the traffic is "out of band" where the "request path" is A -> B, but not A -> B-Mirror.

Basically a Mirror is a live load generator which allows you to get real requests.

[1] https://istio.io/latest/docs/tasks/traffic-management/mirroring/

jshaughn commented:
> How are people reviewing? Pulling the PR and building locally?

@lucasponce  I did a `> hugo serve` locally

lucasponce commented on content/documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc:
```diff
@@ -0,0 +1,58 @@
+
+A typical flow consists of the following steps:
+
+. A portal queries the travels service for available destinations.
```
I guess I used italic for the "travels" service instead.

lucasponce commented on content/documentation/staging/tutorial/02-install-travels-demo/02-design-travels-demo.adoc:
```diff
@@ -0,0 +1,58 @@
+
+* Allow to change the settings of every travel shop simulator (traffic ratio, device, user and type of travel).
+* Providing a *business* view of the total requests generated from the *travel-portal* namespace to the *travel-agency* services, organized with business criteria as grouped per shop, per type of traffic and per city.
+
```
Not sure if I understand your comment here.
It's describing the key features of the *control* app.

jshaughn replied:
I am just saying that at this point, because I had already deployed the travels demo, I thought I was being asked to actually bring up the console/business view.  Then I realized it was just telling me about the console and there was no action to take at this point.   I will resolve this conversation and instead add some specific change recommendations.

jshaughn replied:
Actually, you can just leave it as is.  I'm not sure why I was confused here.   In the overall context it seems fairly clear.  For some reason I was thinking we needed a line prior to the screenshot like:

`Later in the tutorial we will learn how to bring up the console, it looks like this:`

But I no longer think an addition is really necessary.

lucasponce commented on content/documentation/staging/tutorial/03-first-steps/02-enable-sidecars.adoc:
```diff
@@ -0,0 +1,33 @@
+
+In this tutorial, we will add namespaces and workloads into the ServiceMesh individually step by step.
+
+This will help users to understand how Istio sidecar proxies work and how applications can use Istio's features.
```
As tutorial is long, there are some styles where it describes step in third person, and others directly.
But I'm +1 with the change.

lucasponce commented on content/documentation/staging/tutorial/04-observe/03-details.adoc:
```diff
@@ -0,0 +1,103 @@
+
+Next step is to get more context to answer if some particular service is acting slower than expected.
+
+The Traces tab provides a view where traces can be compared with the metrics histogram and user can determine if a particular spike is expected in the context of average values.
```
oh, first time you use the "allows" verb :)

lucasponce commented on content/documentation/staging/tutorial/05-connect/02-fault-injection.adoc:
```diff
@@ -0,0 +1,79 @@
+
+NOTE: *[2]* Understanding *source* and *destination* metrics
+
+Telemetry is collected from proxies and it is labeled with information about the *source* and *destination* calls.
```
"workloads" I think is the term better describe this.

lucasponce commented:
Again, the review work was impressive.
I think I've addressed most of the comments and suggestions.

Let me know how it looks.

Then when staging is ready, I can prepare the copy for latest as well to solve the problems of the missing links.

Thanks

lucasponce commented:
Note that next commit will address the technical issues and it will duplicate the new staging content into latest to comply with the top level menu entries.

So, in case there are any more changes in the content, just comment in the staging area and I'll update them both.

Thanks

lucasponce commented:
Ok, now the top menu "tutorial" link works as it is linked with "latest" folder content:
![image](https://user-images.githubusercontent.com/1662329/105335210-0fbc6300-5bd8-11eb-9f17-353dda230327.png)

lucasponce commented:
Ok, I've fixed the technical issues that make the CI to fail.

To all reviewers: @xeviknal, @jotak, @jshaughn, @jmazzitelli give it a last look before that we exceed the max number of comments allowed per Pull Request :).

Thanks :)

jshaughn commented on content/documentation/staging/tutorial/05-connect/07-mirroring.adoc:
```diff
@@ -0,0 +1,75 @@
+++++
+
+{nbsp} +
+But the "Destination" proxy, in this case, the proxy injected in the *travels-v3* workload, will collect the telemetry from the mirrored traffic.
```
in this case, the --> in this case the

jshaughn commented on content/documentation/staging/tutorial/05-connect/07-mirroring.adoc:
```diff
@@ -0,0 +1,75 @@
+
+NOTE: *Step 3:* Update or delete Istio Configuration
+
+As part of this step you can update the Mirroring scenario to test different mirrored distribution, when finished you can delete the generated Istio config for the *travels* service.
```
distribution,  when  -->  distributions.  When

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+:icons: font
+:sectlinks:
+
+link:https://istio.io/latest/docs/concepts/security/[Security, window="_blank"] is one of the main pillar of Istio features.
```
pillar --> pillars

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+The Istio https://istio.io/latest/docs/concepts/security/#high-level-architecture[Security High Level Architecture] provides a comprehensive solution to design and implement multiple security scenarios.
+
+In the context of this tutorial we are going to show how Kiali can combine telemetry information to create security policies for the workloads deployed on a given namespace.
```
--> In this tutorial we will show how Kiali can use telemetry information to create security policies for the workloads deployed in a given namespace.

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+In the context of this tutorial we are going to show how Kiali can combine telemetry information to create security policies for the workloads deployed on a given namespace.
+
+Istio aggregates in the telemetry the ServiceAccount information used by the workloads in the communication, and this information can be used to define authorization policies with deny and allow actions for a particular live traffic communication status.
```
Split into sentences:

Istio telemetry aggregates the ServiceAccount information used in the workloads communication.  This information can be used to define authorization policies that deny and allow actions on future live traffic communication status.

"on future live traffic communication status."  is a little unclear to me but I don't have a suggestion.

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+Istio aggregates in the telemetry the ServiceAccount information used by the workloads in the communication, and this information can be used to define authorization policies with deny and allow actions for a particular live traffic communication status.
+
+This step will show how we can define authorization policies for the *travel-agency* namespace in the Travel Demo application for all existing traffic in a given time.
```
application for --> application, for
time -->  time period.

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+Once authorization policies are defined, a new workload will be rejected if it doesn't match the security rules defined.
+
+NOTE: *Step 1:* Undeploy the *loadbtester* workload from *travel-portal* namespace
```
loadbtester --> loadtester

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+++++
+
+{nbsp} +
+This will generate a main DENY ALL rule to protect the whole namespace, and individual ALLOW rule per workload identified in the telemetry.
```
and individual --> and an individual

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+NOTE: *Step 3:* Deploy the *loadtester* portal in the *travel-portal* namespace
+
+The *loadtester* workload uses a different ServiceAccount then when it's deployed it won't comply the AuthorizationPolicy rules defined in the previous step.
```
-->  If the *loadtester* uses a different ServiceAccount then, when it's deployed,  it won't comply with the...

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+++++
+
+{nbsp} +
+This can be also verified in the details page using the Outbound Metrics tab grouped by response code.
```
can be also --> can also be

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+++++
+
+{nbsp} +
+Inspecting the Logs tab confirms that *loadtester* workload is getting a HTTP 403 Forbidden response from *travels* workloads as expected.
```
workloads as expected --> workloads, as expected.

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+++++
+
+{nbsp} +
+As expected, now we can see that *travels-v1* workload accepts request from all *travel-portal* namespace workloads but *travels-v2* and *travels-v3* continue rejecting requests from *loadtester* source.
```
request from --> requests from

workloads but --> workloads, but

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+++++
+
+{nbsp} +
+Using "Outbound Metrics" tab from the *loadtester* workload we can group per "Remove version" and "Response code" to get a detailed view of this AuthorizationPolicy change.
```
Remove --> Remote

jshaughn commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+NOTE: *Step 5:* Update or delete Istio Configuration
+
+As part of this step you can update the AuthorizationPolicies generated for the *travel-agency* namespace and experiment with more security rules, or you can delete the generated Istio config for the namespace.
```
namespace and --> namespace, and

lucasponce commented:
> Also, for me all of the links are showing window="_blank" on screen, I think there is just a consistent error to globally fix. The links are not opening the new tab.

Really bad, I guess it's due the hugo version.
Locally I didn't see it:

![image](https://user-images.githubusercontent.com/1662329/105385640-dc97c500-5c13-11eb-8b19-4b8dc704967e.png)

But yes, I can see the problem in the netlify preview link

jshaughn reviewed (commented):
@lucasponce OK, great work!  I am approving this initial version.   One non-blocking comment  is that I think the look and flow of the tutorial would be improved if the screenshots were smaller in their embedded form.  Some whitespace framing would make it much more obvious where the tutorial text ends or begins.  As it is now the text too often flows into the screenshot and screenshots take up a massive amount of vertical space, often forcing a scroll.

I realize that we use the same CSS in other places, like in Features, but I would have no problem if the change was global.  Of course the expanded form would stay the same.

jmazzitelli commented on content/documentation/staging/tutorial/06-secure/01-traffic-policies.adoc:
```diff
@@ -0,0 +1,128 @@
+
+Every workload in the cluster uses a https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/[Service Account, window="_blank"].
+
+*travels.uk*, *viaggi.it* and *voyages.fr* workloads use the default *cluster.local/ns/travel-portal/sa/default" ServiceAccount defined automatically per namespace.
```
I think `*` is meant to be a wildcard here. So I think this is just missing the left-quote:
```suggestion
*travels.uk*, *viaggi.it* and *voyages.fr* workloads use the default "*cluster.local/ns/travel-portal/sa/default" ServiceAccount defined automatically per namespace.
```

jmazzitelli reviewed (changes requested):
This is minor but - the URLs to some of these pages can be confusing. For example, in the first part "Prerequisites" there is the page "Access the Kiali UI" whose URL ends with `tutorial/#04-access-kiali`:
![image](https://user-images.githubusercontent.com/2029470/105427023-4024f100-5c1a-11eb-9d61-d2070bd538f3.png)
But when you go to a later section, e.g. "Observe", and you go to the "Enable Sidecars in all workloads" link, it ends with `tutorial/#01-enable-sidecars-for-all-namespaces`.
![image](https://user-images.githubusercontent.com/2029470/105427262-bcb7cf80-5c1a-11eb-8aa6-f59239df44b4.png)
It just seems weird that an earlier section is number 4 and a later section is number 1. The numbering seems off (I understand the numbering resets for each main subject, but the URLs could be confusing if you don't notice this).

In addition to the above comments, see below for other comments in this review.

lucasponce commented:
> The numbering seems off (I understand the numbering resets for each main subject, but the URLs could be confusing if you don't notice this).

Yes, I see. But those are autogenerated by Hugo, if they are not broken I'd leave them as it is, otherwise we'd need to revisit how Hugo generates the indexing and provide other way.

lucasponce commented on content/documentation/staging/tutorial/01-prerequisites/02-install-istio.adoc:
```diff
@@ -0,0 +1,15 @@
+
+icon:bullhorn[size=1x]{nbsp} Determining ingress IP and ports and creating DNS entries will be necessary in the following steps.
+
+icon:bullhorn[size=1x]{nbsp} This tutorial uses https://istio.io/latest/docs/setup/platform-setup/minikube/[Minikube tunnel, window="_blank"] feature for external Ingress IP.
```
Right. Minikube's tunnel is used to setup an External IP to the Ingress route as part of the prerequisites for the cluster setup.
Otherwise you can link a DNS entry with a public IP.
As this is platform dependend and it may vary in other environments I didn't enter in too many details but +1 to add a note.

lucasponce commented on content/documentation/staging/tutorial/03-first-steps/03-open-application-to-outside.adoc:
```diff
@@ -0,0 +1,97 @@
+
+icon:bullhorn[size=1x]{nbsp} This step depends on the platform you have chosen, please review https://istio.io/latest/docs/setup/getting-started/#determining-the-ingress-ip-and-ports[Determining the Ingress IP and Ports, window="_blank"] for more details.
+
+icon:bullhorn[size=1x]{nbsp} There are multiple ways to create a DNS entry depending of the platform, servers or services that you are using.
```
+1

I put a note like

```
icon:bullhorn[size=1x]{nbsp} Note that EXTERNAL-IP for LoadBalancer are provided in minikube using minikube tunnel tool. As commented, this depends on the platform you are using, review your Istio setup in order to validate the external IP and ports used for Ingress.
```

To clarify that but yes, the install and setup of Istio is out of the scope of the Kiali Tutorial, so I wouldn't enter in too deep details, just mention that it needs the minikube tunnel to get the EXTERNAL-IP.

jmazzitelli replied:
That note would be perfect.

lucasponce commented on content/documentation/staging/tutorial/04-observe/01-enable-sidecars-for-all-namespaces.adoc:
```diff
@@ -0,0 +1,59 @@
+kubectl delete -f <(curl -L https://raw.githubusercontent.com/kiali/demos/master/travels/travel_portal.yaml) -n travel-portal
+
+kubectl apply -f <(curl -L https://raw.githubusercontent.com/kiali/demos/master/travels/travel_agency.yaml) -n travel-agency
+kubectl apply -f <(curl -L https://raw.githubusercontent.com/kiali/demos/master/travels/travel_portal.yaml) -n travel-portal
```
Thanks for the pointer, is this error consistent ?

I mean, as discussed in other comments, in order to enable the istio-injection using the namespace label, the previous controllers need to be deleted and the easiest way could be to perform a delete.

Perhaps a race condition is happening and it is needed to wait the old controllers are deleted, before to install them again ?

lucasponce replied:
I leave a comment in the steps, any other suggestion is welcome.

jmazzitelli replied:
I'm going to try this tutorial again soon, and see what happens.

lucasponce commented on content/documentation/staging/tutorial/05-connect/06-circuit-breaking.adoc:
```diff
@@ -0,0 +1,95 @@
+
+++++
+<a class="image-popup-fit-height" href="/images/tutorial/05-06-loadtester-graph.png" title="Loadtester Graph">
+    <img src="/images/tutorial/05-06-loadtester-graph.png" style="display:block;margin: 0 auto;" />
```
Good catch.
I forgot to commit the images on this section.

lucasponce commented on content/documentation/staging/tutorial/05-connect/06-circuit-breaking.adoc:
```diff
@@ -0,0 +1,95 @@
+
+++++
+<a class="image-popup-fit-height" href="/images/tutorial/05-06-loadtester-details.png" title="Loadtester Outbound Metrics">
+    <img src="/images/tutorial/05-06-loadtester-details.png" style="display:block;margin: 0 auto;" />
```
Good catch.
The good images was missed here.

lucasponce commented:
Thanks for all your suggestions and bug catches.
I think I've addressed most of them, let me know if you miss something.

lucasponce commented:
> One non-blocking comment is that I think the look and flow of the tutorial would be improved if the screenshots were smaller in their embedded form. Some whitespace framing would make it much more obvious where the tutorial text ends or begins. As it is now the text too often flows into the screenshot and screenshots take up a massive amount of vertical space, often forcing a scroll.

> I realize that we use the same CSS in other places, like in Features, but I would have no problem if the change was global. Of course the expanded form would stay the same.

+1 I'm on the same page here.

I don't have any strong suggestion for this, my only comment is to ensure that style is consistent when adding screenshots.

I saw that some pages differ and probably if we want to apply any change on CSS, we'd like to change it for all.

Also, to make easier to write content, and just to put screenshots, links, videos, etc, in the same way and avoid writer to touch the html/css template.

lucasponce commented:
Thanks. With the release of v1.29, I need to rebase/update the PR to support the staging and latest with the new version.

lucasponce commented:
@jmazzitelli @jshaughn @jotak @xeviknal 

I've rebased with master, updated tutorial content with staging and v1.29.0.

Note that, I haven't completely tested the tutorial with v1.29.

There are problem with helm-charts and I used a custom installer, but just to test that all the content is good.

Potential changes in screenshots with v1.29 are not covered, this PR is big, I think it's better to consolidate now and fix things in an incremental way in future steps.

jmazzitelli reviewed (approved):
i'm OK with this to merge.. I've still have yet to be able to follow the steps and get it all to work (I can't do the delete/re-apply step - the pods fail to restart successfully), but I'll assume its a problem on my environment since it appears it works for everyone else

jotak reviewed (approved):
LGTM - not gone through the whole tutorial yet, but it overall looks great and we can still bring later modifications if necessary

xeviknal reviewed (approved):
Great work. Awesome collaborative PR review. Agree to cover "bugs" in subsequent PRs.

jshaughn reviewed (approved):
Great work, let's get thisi initial version in!

lucasponce commented:
Thanks !
